src/common/src/types/ordered.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! `ScalarImpl` and `Datum` wrappers that implement `PartialOrd` and `Ord` with default order type.
+
+use std::cmp::{Ord, Ordering};
+use std::ops::Deref;
+
+use paste::paste;
+use risingwave_common_proc_macro::EstimateSize;
+
+use crate::estimate_size::EstimateSize;
+use crate::for_all_scalar_variants;
+use crate::types::{Datum, DatumRef, ScalarImpl, ScalarRefImpl};
+use crate::util::sort_util::{cmp_datum, partial_cmp_datum, OrderType};
+
+macro_rules! gen_default_partial_cmp_scalar_ref_impl {
+    ($( { $variant_name:ident, $suffix_name:ident, $scalar:ty, $scalar_ref:ty } ),*) => {
+        pub fn default_partial_cmp_scalar_ref_impl(lhs: ScalarRefImpl<'_>, rhs: ScalarRefImpl<'_>) -> Option<Ordering> {
+            match (lhs, rhs) {
+                $((ScalarRefImpl::$variant_name(lhs_inner), ScalarRefImpl::$variant_name(ref rhs_inner)) => Some(lhs_inner.cmp(rhs_inner)),)*
+                _ => None,
+            }
+        }
+    };
+}
+for_all_scalar_variants!(gen_default_partial_cmp_scalar_ref_impl);
+
+/// Variant of [`PartialOrd`] that compares with default order.
+pub trait DefaultPartialOrd {
```

Comment:
I guess we can omit `PartialOrd` for a self-defined `Ord` trait. Currently `PartialOrd` returns `None` only if they're in different types, which is not expected and simply panicking is okay there.